[BUG] `clone!` loses closure arguments span information, leading to poor diagnostics

**Bug description**

If you provide a closure to the `clone!` macro that takes the wrong number of arguments, you'll get a diagnostic like this:

```
error[E0593]: closure is expected to take 4 arguments, but it takes 0 arguments
   --> src/places_sidebar.rs:301:25
    |
301 |               drop_target.connect_drop(clone!(@weak item => move || {
    |  _________________________^^^^^^^^^^^^_-
    | |                         |
    | |                         expected closure that takes 4 arguments
302 | |             }));
    | |______________- takes 0 arguments
    |
help: consider changing the closure to take and ignore the expected arguments
    |
301 |             drop_target.connect_drop(|_, _, _, _|);
    | 
```

This is because rustc is unable to determine the span of the closure arguments, so it falls back to the span of the entire closure. I believe we should be able to thread the correct span back to rustc if we emit the expanded closure arguments with the same span as the original closure. That would lead to a diagnostic like this:

```
error[E0593]: closure is expected to take 4 arguments, but it takes 0 arguments
   --> src/places_sidebar.rs:301:25
    |
301 |               drop_target.connect_drop(clone!(@weak item => move || {
    |                           ^^^^^^^^^^^^                           -- takes 0 arguments
    |                           |
    |                           expected closure that takes 4 arguments
302 |               }));
    |
    |
help: consider changing the closure to take and ignore the expected arguments
    |
301 |             drop_target.connect_drop(clone!(@weak item => move |_, _, _, _| {;
    | 
```

**Backtrace**

N/A

--- a/glib-macros/src/clone.rs
+++ b/glib-macros/src/clone.rs
@@ -3,11 +3,95 @@
 use std::iter::Peekable;
 
 use proc_macro::{
-    token_stream::IntoIter as ProcIter, Delimiter, Group, Ident, Punct, Spacing, Span, TokenStream,
-    TokenTree,
+    token_stream::IntoIter as ProcIter, Delimiter, Group, Ident, Literal, Punct, Spacing, Span,
+    TokenStream, TokenTree,
 };
 
 use crate::utils::crate_ident_new;
+
+struct PeekableProcIter {
+    inner: Peekable<ProcIter>,
+    current_span: Option<Span>,
+    next_span: Option<Span>,
+}
+
+impl<'a> From<&'a Group> for PeekableProcIter {
+    fn from(f: &'a Group) -> Self {
+        let current_span = Some(f.span());
+        let mut inner = f.stream().into_iter().peekable();
+        let next_span = inner.peek().map(|n| n.span());
+        Self {
+            inner,
+            current_span,
+            next_span,
+        }
+    }
+}
+
+impl From<TokenStream> for PeekableProcIter {
+    fn from(f: TokenStream) -> Self {
+        let mut inner = f.into_iter().peekable();
+        let next_span = inner.peek().map(|n| n.span());
+        Self {
+            inner,
+            current_span: None,
+            next_span,
+        }
+    }
+}
+
+impl Iterator for PeekableProcIter {
+    type Item = TokenTree;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.inner.next()?;
+        self.current_span = Some(next.span());
+        self.next_span = self.inner.peek().map(|n| n.span());
+        Some(next)
+    }
+}
+
+impl PeekableProcIter {
+    fn peek(&mut self) -> Option<&TokenTree> {
+        self.inner.peek()
+    }
+
+    fn generate_error(&self, message: &str) -> TokenStream {
+        self.generate_error_with_span(message, self.current_span)
+    }
+
+    fn generate_error_with_next_span(&self, message: &str) -> TokenStream {
+        self.generate_error_with_span(message, self.next_span)
+    }
+
+    fn generate_error_with_span(&self, message: &str, span: Option<Span>) -> TokenStream {
+        let span = span.unwrap_or_else(Span::call_site);
+        // We generate a `compile_error` macro and assign it to the current span so the error
+        // displayed by rustc points to the right location.
+        let mut stream = TokenStream::new();
+        stream.extend(vec![TokenTree::Literal(Literal::string(message))]);
+
+        let mut tokens = vec![
+            TokenTree::Ident(Ident::new("compile_error", span)),
+            TokenTree::Punct(Punct::new('!', Spacing::Alone)),
+            TokenTree::Group(Group::new(Delimiter::Parenthesis, stream)),
+            TokenTree::Punct(Punct::new(';', Spacing::Alone)),
+        ];
+
+        for tok in &mut tokens {
+            tok.set_span(span);
+        }
+
+        let mut stream = TokenStream::new();
+        stream.extend(tokens);
+        let mut t = TokenTree::Group(Group::new(Delimiter::Brace, stream));
+        t.set_span(span);
+
+        let mut stream = TokenStream::new();
+        stream.extend(vec![t]);
+        stream
+    }
+}
 
 #[derive(Clone, Copy, Debug)]
 enum BorrowKind {
@@ -194,19 +278,17 @@ enum TokenCheck {
 
 fn check_tokens(
     tokens_to_check: &[SimpleToken],
-    parts: &mut Peekable<ProcIter>,
+    parts: &mut PeekableProcIter,
 ) -> Result<(), TokenCheck> {
     let mut tokens = String::new();
 
     for token in tokens_to_check {
-        if let Some(next) = parts.next() {
-            if *token != next {
-                return Err(TokenCheck::UnexpectedToken(
-                    tokens,
-                    token.to_str().to_owned(),
-                ));
+        if let Some(next) = parts.peek() {
+            if token != next {
+                return Err(TokenCheck::UnexpectedToken(tokens, next.to_string()));
             }
             tokens.push_str(token.to_str());
+            parts.next();
         } else {
             return Err(TokenCheck::UnexpectedEnd(tokens));
         }
@@ -215,7 +297,10 @@ fn check_tokens(
 }
 
 #[doc(alias = "get_full_ident")]
-fn full_ident(parts: &mut Peekable<ProcIter>, borrow_kind: BorrowKind) -> String {
+fn full_ident(
+    parts: &mut PeekableProcIter,
+    borrow_kind: BorrowKind,
+) -> Result<String, TokenStream> {
     let mut name = String::new();
     let mut prev_is_ident = false;
 
@@ -226,18 +311,22 @@ fn full_ident(parts: &mut Peekable<ProcIter>, borrow_kind: BorrowKind) -> String
                 if p_s == "," || p_s == "=" {
                     break;
                 } else if p_s == "." {
-                    assert!(
-                        prev_is_ident,
-                        "Unexpected `.` after `{}`",
-                        borrow_kind.to_str()
-                    );
+                    if !prev_is_ident {
+                        return Err(parts.generate_error_with_next_span(&format!(
+                            "Unexpected `.` after `{}`",
+                            borrow_kind.to_str()
+                        )));
+                    }
                     prev_is_ident = false;
                     name.push('.');
                     parts.next();
                 } else if name.is_empty() {
-                    panic!("Expected ident, found `{p_s}`");
+                    return Err(parts
+                        .generate_error_with_next_span(&format!("Expected ident, found `{p_s}`")));
                 } else {
-                    panic!("Expected ident, found `{p_s}` after `{name}`");
+                    return Err(parts.generate_error_with_next_span(&format!(
+                        "Expected ident, found `{p_s}` after `{name}`"
+                    )));
                 }
             }
             Some(TokenTree::Ident(i)) => {
@@ -248,24 +337,36 @@ fn full_ident(parts: &mut Peekable<ProcIter>, borrow_kind: BorrowKind) -> String
                 name.push_str(&i.to_string());
                 parts.next();
             }
-            Some(x) if name.is_empty() => panic!("Expected ident, found `{x}`"),
-            Some(x) => panic!("Expected ident, found `{x}` after `{name}`"),
-            None => panic!("Unexpected end after ident `{name}`"),
+            Some(x) if name.is_empty() => {
+                let err = format!("Expected ident, found `{x}`");
+                return Err(parts.generate_error_with_next_span(&err));
+            }
+            Some(x) => {
+                let err = &format!("Expected ident, found `{x}` after `{name}`");
+                return Err(parts.generate_error_with_next_span(err));
+            }
+            None => {
+                return Err(parts.generate_error(&format!("Unexpected end after ident `{name}`")));
+            }
         }
     }
-    assert!(
-        !name.is_empty(),
-        "Expected ident, found `{}`",
-        parts.next().unwrap()
-    );
-    name
+    if name.is_empty() {
+        if let Some(next) = parts.next() {
+            return Err(parts.generate_error(&format!("Expected ident, found `{next}`")));
+        }
+        return Err(parts.generate_error("Expected something after, found nothing"));
+    }
+    Ok(name)
 }
 
 #[doc(alias = "get_keyword")]
-fn keyword(parts: &mut Peekable<ProcIter>) -> String {
+fn keyword(parts: &mut PeekableProcIter) -> Result<BorrowKind, TokenStream> {
     let mut ret = String::new();
     let mut prev_is_ident = false;
     let mut stored = false;
+    // We unfortunately can't join spans since the `Span::join` method is nightly-only. Well, we'll
+    // do our best...
+    let start_span = parts.next_span;
 
     loop {
         match parts.peek() {
@@ -292,51 +393,86 @@ fn keyword(parts: &mut Peekable<ProcIter>) -> String {
         }
         parts.next();
     }
-    ret
-}
-
-fn parse_ident(parts: &mut Peekable<ProcIter>, elements: &mut Vec<ElemToClone>) {
-    let borrow_kind = match keyword(parts).as_str() {
+    let ret = match ret.as_str() {
         "strong" => BorrowKind::Strong,
         "weak" => BorrowKind::Weak,
         "weak-allow-none" => BorrowKind::WeakAllowNone,
         "to-owned" => BorrowKind::ToOwned,
-        "default-return" => panic!("`@default-return` should be after `=>`"),
-        "default-panic" => panic!("`@default-panic` should be after `=>`"),
-        k => panic!(
-            "Unknown keyword `{k}`, only `weak`, `weak-allow-none`, `to-owned` and `strong` are allowed",
-        ),
+        "default-return" => {
+            return Err(parts
+                .generate_error_with_span("`@default-return` should be after `=>`", start_span));
+        }
+        "default-panic" => {
+            return Err(
+                parts.generate_error_with_span("`@default-panic` should be after `=>`", start_span)
+            );
+        }
+        k => {
+            return Err(parts.generate_error_with_span(
+                &format!(
+                    "Unknown keyword `{k}`, only `weak`, `weak-allow-none`, `to-owned` and \
+                    `strong` are allowed"
+                ),
+                start_span,
+            ));
+        }
     };
-    let name = full_ident(parts, borrow_kind);
+    Ok(ret)
+}
+
+fn parse_ident(
+    parts: &mut PeekableProcIter,
+    elements: &mut Vec<ElemToClone>,
+) -> Result<(), TokenStream> {
+    let borrow_kind = keyword(parts)?;
+    let name = full_ident(parts, borrow_kind)?;
+    let name_span = parts.current_span;
+    if name.ends_with('.') {
+        return Err(
+            parts.generate_error_with_span(&format!("Invalid variable name: `{name}`"), name_span)
+        );
+    }
     let alias = match parts.peek() {
         Some(TokenTree::Ident(p)) if p.to_string() == "as" => {
             parts.next();
+            let current_span = parts.current_span;
             match parts.next() {
                 Some(TokenTree::Ident(i)) => Some(i.to_string()),
-                Some(x) => panic!("Expected ident after `as` keyword, found `{x}`"),
-                None => panic!("Unexpected end after `as` keyword"),
+                Some(x) => {
+                    let err = format!("Expected ident after `as` keyword, found `{x}`");
+                    return Err(parts.generate_error(&err));
+                }
+                None => {
+                    return Err(parts.generate_error_with_span(
+                        "Unexpected end after `as` keyword",
+                        current_span,
+                    ))
+                }
             }
         }
-        Some(TokenTree::Ident(p)) => panic!("Unexpected `{p}`"),
+        Some(TokenTree::Ident(p)) => {
+            let err = format!("Unexpected `{p}`");
+            return Err(parts.generate_error(&err));
+        }
         _ => None,
     };
     if name == "self" && alias.is_none() {
-        panic!(
+        return Err(parts.generate_error_with_span(
             "Can't use `self` as variable name. Try storing it in a temporary variable or \
-                rename it using `as`."
-        );
-    } else {
-        assert!(!name.ends_with('.'), "Invalid variable name: `{name}`");
-        assert!(
-            !name.contains('.') || alias.is_some(),
-            "`{name}`: Field accesses are not allowed as is, you must rename it!",
-        );
+                rename it using `as`.",
+            name_span,
+        ));
+    } else if name.contains('.') && alias.is_none() {
+        let err = format!("`{name}`: Field accesses are not allowed as is, you must rename it!");
+        return Err(parts.generate_error_with_span(&err, name_span));
     }
+
     elements.push(ElemToClone {
         name,
         alias,
         borrow_kind,
     });
+    Ok(())
 }
 
 fn delimiter_to_string(delimiter: Delimiter, open: bool) -> &'static str {
@@ -370,15 +506,16 @@ fn group_to_string(g: &Group) -> String {
     format!(
         "{}{}{}",
         delimiter_to_string(g.delimiter(), true),
-        tokens_to_string(g.stream().into_iter().peekable()),
+        tokens_to_string(PeekableProcIter::from(g)),
         delimiter_to_string(g.delimiter(), false),
     )
 }
 
 #[doc(alias = "get_expr")]
-fn expr(parts: &mut Peekable<ProcIter>) -> String {
+fn expr(parts: &mut PeekableProcIter) -> Result<String, TokenStream> {
     let mut ret = String::new();
     let mut total = 0;
+    let span = parts.current_span;
     match parts.next() {
         Some(TokenTree::Literal(l)) => ret.push_str(&l.to_string()),
         Some(TokenTree::Ident(i)) => ret.push_str(&i.to_string()),
@@ -386,10 +523,17 @@ fn expr(parts: &mut Peekable<ProcIter>) -> String {
             "[" | "{" | "(" => {
                 total += 1;
             }
-            x => panic!("Unexpected token `{x}` after `@default-return`"),
+            x => {
+                return Err(parts
+                    .generate_error(&format!("Unexpected token `{x}` after `@default-return`")))
+            }
         },
-        Some(TokenTree::Group(g)) => return group_to_string(&g),
-        None => panic!("Unexpected end after `@default-return`"),
+        Some(TokenTree::Group(g)) => return Ok(group_to_string(&g)),
+        None => {
+            return Err(
+                parts.generate_error_with_span("Unexpected end after `@default-return`", span)
+            )
+        }
     };
     loop {
         match parts.peek() {
@@ -400,7 +544,7 @@ fn expr(parts: &mut Peekable<ProcIter>) -> String {
                 } else if p_s == "}" || p_s == ")" || p_s == "]" || p_s == ">" {
                     total -= 1;
                 } else if p_s == "," && total == 0 {
-                    return ret;
+                    return Ok(ret);
                 }
                 ret.push_str(&p_s);
             }
@@ -409,67 +553,77 @@ fn expr(parts: &mut Peekable<ProcIter>) -> String {
             }
             Some(x) => {
                 if total == 0 && !ret.ends_with(':') {
-                    return ret;
+                    return Ok(ret);
                 }
                 ret.push_str(&x.to_string())
             }
-            None => panic!(
+            None => return Err(parts.generate_error(
                 "Unexpected end after `{ret}`. Did you forget a `,` after the @default-return value?",
-            ),
+            )),
         }
         parts.next();
     }
 }
 
 #[doc(alias = "get_return_kind")]
-fn return_kind(parts: &mut Peekable<ProcIter>) -> WrapperKind {
+fn return_kind(parts: &mut PeekableProcIter) -> Result<WrapperKind, TokenStream> {
     match check_tokens(
         &[SimpleToken::Ident("default"), SimpleToken::Punct("-")],
         parts,
     ) {
         Err(TokenCheck::UnexpectedToken(tokens, unexpected_token)) => {
-            panic!("Unknown keyword `{tokens}{unexpected_token}`");
+            return Err(
+                parts.generate_error(&format!("Unknown keyword `{tokens}{unexpected_token}`"))
+            );
         }
         Err(TokenCheck::UnexpectedEnd(tokens)) => {
-            panic!("Unexpected end after tokens `{tokens}`");
+            return Err(parts.generate_error(&format!("Unexpected end after tokens `{tokens}`")));
         }
         Ok(()) => {}
     }
+    let prev = parts.current_span;
     match parts.next() {
         Some(TokenTree::Ident(i)) => {
             let i_s = i.to_string();
             if i_s == "panic" {
-                return WrapperKind::DefaultPanic;
+                return Ok(WrapperKind::DefaultPanic);
             }
             assert!(i_s == "return", "Unknown keyword `@default-{i_s}`");
         }
-        Some(x) => panic!("Unknown token `{x}` after `@default-`",),
-        None => panic!("Unexpected end after `@default-`"),
+        Some(x) => {
+            let err = format!("Unknown token `{x}` after `@default-`");
+            return Err(parts.generate_error(&err));
+        }
+        None => {
+            return Err(parts.generate_error_with_span("Unexpected end after `@default-`", prev))
+        }
     }
-    WrapperKind::DefaultReturn(expr(parts))
+    Ok(WrapperKind::DefaultReturn(expr(parts)?))
 }
 
-fn parse_return_kind(parts: &mut Peekable<ProcIter>) -> Option<WrapperKind> {
+fn parse_return_kind(parts: &mut PeekableProcIter) -> Result<Option<WrapperKind>, TokenStream> {
     match parts.peek() {
         Some(TokenTree::Punct(p)) if p.to_string() == "@" => {}
-        None => panic!("Unexpected end 2"),
-        _ => return None,
+        None => return Err(parts.generate_error("Unexpected end 2")),
+        _ => return Ok(None),
     }
     parts.next();
-    let ret = return_kind(parts);
+    let ret = return_kind(parts)?;
     match check_tokens(&[SimpleToken::Punct(",")], parts) {
         Err(TokenCheck::UnexpectedToken(_, unexpected_token)) => {
-            panic!(
+            let err = format!(
                 "Expected `,` after `{}`, found `{unexpected_token}`",
                 ret.to_str(),
             );
+            return Err(parts.generate_error_with_next_span(&err));
         }
         Err(TokenCheck::UnexpectedEnd(tokens)) => {
-            panic!("Expected `,` after `{}{tokens}`", ret.to_str());
+            let err = format!("Expected `,` after `{}{tokens}`", ret.to_str());
+            return Err(parts.generate_error(&err));
         }
         Ok(()) => {}
     }
-    Some(ret)
+    Ok(Some(ret))
 }
 
 enum BlockKind {
@@ -489,108 +643,132 @@ impl BlockKind {
     }
 }
 
-fn check_move_after_async(parts: &mut Peekable<ProcIter>) {
+fn check_move_after_async(parts: &mut PeekableProcIter) -> Result<(), TokenStream> {
+    let span = parts.current_span;
     match parts.next() {
-        Some(TokenTree::Ident(i)) if i.to_string() == "move" => {}
+        Some(TokenTree::Ident(i)) if i.to_string() == "move" => Ok(()),
         // The next checks are just for better error messages.
         Some(TokenTree::Ident(i)) => {
-            panic!("Expected `move` after `async`, found `{i}`");
+            let err = format!("Expected `move` after `async`, found `{i}`");
+            Err(parts.generate_error(&err))
         }
         Some(TokenTree::Punct(p)) => {
-            panic!("Expected `move` after `async`, found `{p}`");
+            let err = format!("Expected `move` after `async`, found `{p}`");
+            Err(parts.generate_error(&err))
         }
         Some(TokenTree::Group(g)) => {
-            panic!(
+            let err = format!(
                 "Expected `move` after `async`, found `{}`",
                 delimiter_to_string(g.delimiter(), true),
             );
+            Err(parts.generate_error(&err))
         }
-        _ => panic!("Expected `move` after `async`"),
+        _ => Err(parts.generate_error_with_span("Expected `move` after `async`", span)),
     }
 }
 
-fn check_async_syntax(parts: &mut Peekable<ProcIter>) -> BlockKind {
-    check_move_after_async(parts);
+fn check_async_syntax(parts: &mut PeekableProcIter) -> Result<BlockKind, TokenStream> {
+    check_move_after_async(parts)?;
     match parts.peek() {
         Some(TokenTree::Punct(p)) if p.to_string() == "|" => {
             parts.next();
-            BlockKind::AsyncClosure(closure(parts))
+            Ok(BlockKind::AsyncClosure(closure(parts)?))
         }
         Some(TokenTree::Punct(p)) => {
-            panic!("Expected closure or block after `async move`, found `{p}`",);
+            let err = format!("Expected closure or block after `async move`, found `{p}`");
+            Err(parts.generate_error_with_next_span(&err))
         }
-        Some(TokenTree::Group(g)) if g.delimiter() == Delimiter::Brace => BlockKind::AsyncBlock,
+        Some(TokenTree::Group(g)) if g.delimiter() == Delimiter::Brace => Ok(BlockKind::AsyncBlock),
         Some(TokenTree::Group(g)) => {
-            panic!(
+            let err = format!(
                 "Expected closure or block after `async move`, found `{}`",
                 delimiter_to_string(g.delimiter(), true),
             );
+            Err(parts.generate_error_with_next_span(&err))
         }
-        _ => panic!("Expected closure or block after `async move`"),
+        _ => Err(parts.generate_error("Expected closure or block after `async move`")),
     }
 }
 
 // Returns `true` if this is an async context.
-fn check_before_closure(parts: &mut Peekable<ProcIter>) -> BlockKind {
+fn check_before_closure(parts: &mut PeekableProcIter) -> Result<BlockKind, TokenStream> {
     let is_async = match parts.peek() {
         Some(TokenTree::Ident(i)) if i.to_string() == "move" => false,
         Some(TokenTree::Ident(i)) if i.to_string() == "async" => true,
         Some(TokenTree::Ident(i)) if i.to_string() == "default" => {
-            let ret = return_kind(parts);
-            panic!("Missing `@` before `{}`", ret.keyword());
+            let span = parts.next_span;
+            let ret = return_kind(parts)?;
+            let err = format!("Missing `@` before `{}`", ret.keyword());
+            return Err(parts.generate_error_with_span(&err, span));
         }
         Some(TokenTree::Punct(p)) if p.to_string() == "|" => {
-            panic!("Closure needs to be \"moved\" so please add `move` before closure")
+            return Err(parts.generate_error_with_next_span(
+                "Closure needs to be \"moved\" so please add `move` before closure",
+            ));
         }
-        _ => panic!("Missing `move` and closure declaration"),
+        _ => {
+            return Err(
+                parts.generate_error_with_next_span("Missing `move` and closure declaration")
+            )
+        }
     };
     parts.next();
     if is_async {
         return check_async_syntax(parts);
     }
-    match parts.next() {
+    match parts.peek() {
         Some(TokenTree::Punct(p)) if p.to_string() == "|" => {}
-        Some(x) => panic!("Expected closure, found `{x}`"),
-        None => panic!("Expected closure"),
+        Some(x) => {
+            let err = format!("Expected closure, found `{x}`");
+            return Err(parts.generate_error_with_next_span(&err));
+        }
+        None => return Err(parts.generate_error("Expected closure")),
     }
-    let closure = closure(parts);
+    parts.next();
+
+    let closure = closure(parts)?;
     match parts.peek() {
         Some(TokenTree::Ident(i)) if i.to_string() == "async" => {
             parts.next();
-            check_move_after_async(parts);
+            check_move_after_async(parts)?;
             match parts.peek() {
                 Some(TokenTree::Group(g)) if g.delimiter() == Delimiter::Brace => {
-                    BlockKind::ClosureWrappingAsync(closure)
+                    Ok(BlockKind::ClosureWrappingAsync(closure))
                 }
                 // The next matchings are for better error messages.
                 Some(TokenTree::Punct(p)) => {
-                    panic!("Expected block after `| async move`, found `{p}`");
+                    let err = format!("Expected block after `| async move`, found `{p}`");
+                    Err(parts.generate_error_with_next_span(&err))
                 }
                 Some(TokenTree::Group(g)) => {
-                    panic!(
+                    let err = format!(
                         "Expected block after `| async move`, found `{}`",
                         delimiter_to_string(g.delimiter(), true),
                     );
+                    Err(parts.generate_error_with_next_span(&err))
                 }
-                _ => panic!("Expected block after `| async move`"),
+                _ => {
+                    Err(parts.generate_error_with_next_span("Expected block after `| async move`"))
+                }
             }
         }
-        _ => BlockKind::Closure(closure),
+        _ => Ok(BlockKind::Closure(closure)),
     }
 }
 
 #[doc(alias = "get_closure")]
-fn closure(parts: &mut Peekable<ProcIter>) -> Vec<TokenTree> {
+fn closure(parts: &mut PeekableProcIter) -> Result<Vec<TokenTree>, TokenStream> {
     let mut ret = Vec::new();
 
     loop {
+        let span = parts.current_span;
         match parts.next() {
             Some(TokenTree::Punct(p)) if p.to_string() == "|" => break,
             Some(x) => ret.push(x),
-            None => panic!("Unexpected end 3"),
+            None => return Err(parts.generate_error_with_span("Unexpected end 3", span)),
         }
     }
-    ret
+    Ok(ret)
 }
 
 pub fn tokens_to_string(parts: impl Iterator<Item = TokenTree>) -> String {
@@ -623,7 +801,7 @@ pub fn tokens_to_string(parts: impl Iterator<Item = TokenTree>) -> String {
 }
 
 fn build_closure(
-    parts: Peekable<ProcIter>,
+    parts: PeekableProcIter,
     elements: Vec<ElemToClone>,
     return_kind: Option<WrapperKind>,
     kind: BlockKind,
@@ -715,11 +893,12 @@ fn build_closure(
 }
 
 pub(crate) fn clone_inner(item: TokenStream) -> TokenStream {
-    let mut parts = item.into_iter().peekable();
+    let mut parts: PeekableProcIter = item.into();
     let mut elements = Vec::new();
     let mut prev_is_ident = false;
 
     loop {
+        let prev = parts.current_span;
         match parts.next() {
             Some(TokenTree::Punct(ref p)) => {
                 let p_s = p.to_string();
@@ -727,7 +906,9 @@ pub(crate) fn clone_inner(item: TokenStream) -> TokenStream {
                     parts.next();
                     break;
                 } else if p_s == "@" {
-                    parse_ident(&mut parts, &mut elements);
+                    if let Err(e) = parse_ident(&mut parts, &mut elements) {
+                        return e;
+                    }
                     prev_is_ident = true;
                 } else if p_s == "," {
                     assert!(prev_is_ident, "Unexpected `,`");
@@ -737,24 +918,34 @@ pub(crate) fn clone_inner(item: TokenStream) -> TokenStream {
                         !elements.is_empty(),
                         "If you have nothing to clone, no need to use this macro!"
                     );
-                    panic!("Expected `=>` before closure");
+                    return parts.generate_error("Expected `=>` before closure");
                 }
             }
             Some(TokenTree::Ident(i)) => {
-                panic!(
+                let err = format!(
                     "Unexpected ident `{i}`: you need to specify if this is a weak or a strong \
                      clone.",
                 );
+                return parts.generate_error(&err);
             }
-            Some(t) => panic!("Unexpected token `{t}`"),
-            None => panic!("Unexpected end 4"),
+            Some(t) => {
+                let err = format!("Unexpected token `{t}`");
+                return parts.generate_error(&err);
+            }
+            None => return parts.generate_error_with_span("Unexpected end 4", prev),
         }
     }
     assert!(
         !elements.is_empty(),
         "If you have nothing to clone, no need to use this macro!"
     );
-    let return_kind = parse_return_kind(&mut parts);
-    let kind = check_before_closure(&mut parts);
+    let return_kind = match parse_return_kind(&mut parts) {
+        Ok(r) => r,
+        Err(e) => return e,
+    };
+    let kind = match check_before_closure(&mut parts) {
+        Ok(r) => r,
+        Err(e) => return e,
+    };
     build_closure(parts, elements, return_kind, kind)
 }

--- a/glib-macros/tests/clone.rs
+++ b/glib-macros/tests/clone.rs
@@ -17,65 +17,242 @@ fn clone() {
 }
 
 const TESTS: &[(&str, &str)] = &[
-    ("clone!( => move || {})",
-     "If you have nothing to clone, no need to use this macro!"),
-    ("clone!(|| {})",
-     "If you have nothing to clone, no need to use this macro!"),
-    ("clone!(|a, b| {})",
-     "If you have nothing to clone, no need to use this macro!"),
-    ("clone!(@weak a, @weak b => |x| {})",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak a, @weak b => || {})",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak a, @weak b => |x| println!(\"a\"))",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak a, @weak b => || println!(\"a\"))",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak a => |x| {})",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak a => || {})",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak a => |x| println!(\"a\"))",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak a => || println!(\"a\"))",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@strong self => move |x| {})",
-     "Can't use `self` as variable name. Try storing it in a temporary variable or rename it using `as`."),
-    ("clone!(@strong self.v => move |x| {})",
-     "`self.v`: Field accesses are not allowed as is, you must rename it!"),
-    ("clone!(@weak v => @default-return false, || {})",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak v => @default-return false, || println!(\"a\"))",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak v => @default-return false, |bla| {})",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak v => @default-return false, |bla| println!(\"a\"))",
-     "Closure needs to be \"moved\" so please add `move` before closure"),
-    ("clone!(@weak v => default-return false, move || {})",
-     "Missing `@` before `default-return`"),
-    ("clone!(@weak v => @default-return false move || {})",
-     "Expected `,` after `@default-return false`, found `,`"),
-    ("clone!(@yolo v => move || {})",
-     "Unknown keyword `yolo`, only `weak`, `weak-allow-none`, `to-owned` and `strong` are allowed"),
-    ("clone!(v => move || {})",
-     "Unexpected ident `v`: you need to specify if this is a weak or a strong clone."),
-    ("clone!(@strong v => {println!(\"foo\");})",
-     "Missing `move` and closure declaration"),
-    ("clone!(@strong v, @default-return lol => move || {println!(\"foo\");})",
-     "`@default-return` should be after `=>`"),
-    ("clone!(@default-return lol, @strong v => move || {println!(\"foo\");})",
-     "`@default-return` should be after `=>`"),
+    (
+        "clone!( => move || {})",
+        "If you have nothing to clone, no need to use this macro!",
+    ),
+    (
+        "clone!(|| {})",
+        "If you have nothing to clone, no need to use this macro!",
+    ),
+    (
+        "clone!(|a, b| {})",
+        "If you have nothing to clone, no need to use this macro!",
+    ),
+    (
+        "clone!(@weak a, @weak b => |x| {})",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_3.rs:1:86
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a, @weak b => |x| {}); }
+  |                                                                                      ^"#,
+    ),
+    (
+        "clone!(@weak a, @weak b => || {})",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_4.rs:1:86
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a, @weak b => || {}); }
+  |                                                                                      ^"#,
+    ),
+    (
+        "clone!(@weak a, @weak b => |x| println!(\"a\"))",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_5.rs:1:86
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a, @weak b => |x| println!("a")); }
+  |                                                                                      ^"#,
+    ),
+    (
+        "clone!(@weak a, @weak b => || println!(\"a\"))",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_6.rs:1:86
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a, @weak b => || println!("a")); }
+  |                                                                                      ^"#,
+    ),
+    (
+        "clone!(@weak a => |x| {})",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_7.rs:1:77
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a => |x| {}); }
+  |                                                                             ^"#,
+    ),
+    (
+        "clone!(@weak a => || {})",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_8.rs:1:77
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a => || {}); }
+  |                                                                             ^"#,
+    ),
+    (
+        "clone!(@weak a => |x| println!(\"a\"))",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_9.rs:1:77
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a => |x| println!("a")); }
+  |                                                                             ^"#,
+    ),
+    (
+        "clone!(@weak a => || println!(\"a\"))",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_10.rs:1:77
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak a => || println!("a")); }
+  |                                                                             ^"#,
+    ),
+    (
+        "clone!(@strong self => move |x| {})",
+        r#"error: Can't use `self` as variable name. Try storing it in a temporary variable or rename it using `as`.
+ --> test_11.rs:1:74
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong self => move |x| {}); }
+  |                                                                          ^^^^
+"#,
+    ),
+    (
+        "clone!(@strong self.v => move |x| {})",
+        r#"error: `self.v`: Field accesses are not allowed as is, you must rename it!
+ --> test_12.rs:1:79
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong self.v => move |x| {}); }
+  |                                                                               ^
+"#,
+    ),
+    (
+        "clone!(@weak v => @default-return false, || {})",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_13.rs:1:100
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak v => @default-return false, || {}); }
+  |                                                                                                    ^"#,
+    ),
+    (
+        "clone!(@weak v => @default-return false, || println!(\"a\"))",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_14.rs:1:100
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak v => @default-return false, || println!("a")); }
+  |                                                                                                    ^"#,
+    ),
+    (
+        "clone!(@weak v => @default-return false, |bla| {})",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_15.rs:1:100
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak v => @default-return false, |bla| {}); }
+  |                                                                                                    ^
+"#,
+    ),
+    (
+        "clone!(@weak v => @default-return false, |bla| println!(\"a\"))",
+        r#"error: Closure needs to be "moved" so please add `move` before closure
+ --> test_16.rs:1:100
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak v => @default-return false, |bla| println!("a")); }
+  |                                                                                                    ^
+"#,
+    ),
+    (
+        "clone!(@weak v => default-return false, move || {})",
+        r#"error: Missing `@` before `default-return`
+ --> test_17.rs:1:77
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak v => default-return false, move || {}); }
+  |                                                                             ^^^^^^^
+"#,
+    ),
+    (
+        "clone!(@weak v => @default-return false move || {})",
+        r#"error: Expected `,` after `@default-return false`, found `move`
+ --> test_18.rs:1:99
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@weak v => @default-return false move || {}); }
+  |                                                                                                   ^^^^
+"#,
+    ),
+    (
+        "clone!(@yolo v => move || {})",
+        r#"error: Unknown keyword `yolo`, only `weak`, `weak-allow-none`, `to-owned` and `strong` are allowed
+ --> test_19.rs:1:67
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@yolo v => move || {}); }
+  |                                                                   ^^^^
+"#,
+    ),
+    (
+        "clone!(v => move || {})",
+        r#"error: Unexpected ident `v`: you need to specify if this is a weak or a strong clone.
+ --> test_20.rs:1:66
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(v => move || {}); }
+  |                                                                  ^
+"#,
+    ),
+    (
+        "clone!(@strong v => {println!(\"foo\");})",
+        r#"error: Missing `move` and closure declaration
+ --> test_21.rs:1:79
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong v => {println!("foo");}); }
+  |                                                                               ^^^^^^^^^^^^^^^^^^
+"#,
+    ),
+    (
+        "clone!(@strong v, @default-return lol => move || {println!(\"foo\");})",
+        r#"error: `@default-return` should be after `=>`
+ --> test_22.rs:1:78
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong v, @default-return lol => move || {println!("foo");}); }
+  |                                                                              ^^^^^^^
+"#,
+    ),
+    (
+        "clone!(@default-return lol, @strong v => move || {println!(\"foo\");})",
+        r#"error: `@default-return` should be after `=>`
+ --> test_23.rs:1:67
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@default-return lol, @strong v => move || {println!("foo");}); }
+  |                                                                   ^^^^^^^
+"#,
+    ),
     // The async part!
-    ("clone!(@strong v => async || {println!(\"foo\");})",
-     "Expected `move` after `async`, found `|`"),
-    ("clone!(@strong v => async {println!(\"foo\");})",
-     "Expected `move` after `async`, found `{`"),
-    ("clone!(@strong v => move || async {println!(\"foo\");})",
-     "Expected `move` after `async`, found `{`"),
-    ("clone!(@strong v => move || async println!(\"foo\");)",
-     "Expected `move` after `async`, found `println`"),
-    ("clone!(@strong v => move || async move println!(\"foo\");)",
-     "Expected block after `| async move`"),
+    (
+        "clone!(@strong v => async || {println!(\"foo\");})",
+        r#"error: Expected `move` after `async`, found `|`
+ --> test_24.rs:1:85
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong v => async || {println!("foo");}); }
+  |                                                                                     ^"#,
+    ),
+    (
+        "clone!(@strong v => async {println!(\"foo\");})",
+        r#"error: Expected `move` after `async`, found `{`
+ --> test_25.rs:1:85
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong v => async {println!("foo");}); }
+  |                                                                                     ^^^^^^^^^^^^^^^^^^
+"#,
+    ),
+    (
+        "clone!(@strong v => move || async {println!(\"foo\");})",
+        r#"error: Expected `move` after `async`, found `{`
+ --> test_26.rs:1:93
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong v => move || async {println!("foo");}); }
+  |                                                                                             ^^^^^^^^^^^^^^^^^^
+"#,
+    ),
+    (
+        "clone!(@strong v => move || async println!(\"foo\");)",
+        r#"error: Expected `move` after `async`, found `println`
+ --> test_27.rs:1:93
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong v => move || async println!("foo");); }
+  |                                                                                             ^^^^^^^
+"#,
+    ),
+    (
+        "clone!(@strong v => move || async move println!(\"foo\");)",
+        r#"error: Expected block after `| async move`
+ --> test_28.rs:1:98
+  |
+1 | fn main() { use glib::clone; let v = std::rc::Rc::new(1); clone!(@strong v => move || async move println!("foo");); }
+  |                                                                                                  ^^^^^^^
+"#,
+    ),
 ];
 
 #[test]

